numba/tests/test_linalg.py

```diff
@@ -1431,5 +1435,129 @@ def check(a, b, **kwargs):
         self.assert_raise_on_singular(cfunc, (bad2D, ok))
 
 
+class TestLinalgPinv(TestLinalgBase):
+    """
+    Tests for np.linalg.pinv.
+    """
+
+    @needs_lapack
+    def test_linalg_pinv(self):
+        """
+        Test np.linalg.pinv
+        """
+        cfunc = jit(nopython=True)(pinv_matrix)
+
+        def check(a, **kwargs):
+            expected = pinv_matrix(a, **kwargs)
+            got = cfunc(a, **kwargs)
+
+            # check that the computed results are contig and in the same way
+            self.assert_contig_sanity(got, "F")
+
+            use_reconstruction = False
+            # try plain match of each array to np first
+
+            try:
+                np.testing.assert_array_almost_equal_nulp(
+                    got, expected, nulp=10)
+            except AssertionError:
+                # plain match failed, test by reconstruction
+                use_reconstruction = True
+
+            # If plain match fails then reconstruction is used.
+            # This can occur due to numpy using double precision
+            # LAPACK when single can be used, this creates round off
+            # problems. Also, if the matrix has machine precision level
+            # zeros in its singular values then the singular vectors are 
+            # likely to vary depending on round off.
+            if use_reconstruction:
+
+                # check they are dimensionally correct
+                self.assertEqual(got.shape, expected.shape)
+
+                # check pinv(A)*A~=eye
+                # if the problem is numerical fuzz then this will probably
+                # work, if the problem is rank deficiency then it won't!
+                rec = np.dot(got, a)
+                eye = np.eye(a.shape[-1], dtype=a.dtype)
+                resolution = 5 * np.finfo(a.dtype).resolution
+                try:
+                    np.testing.assert_allclose(
```

Comment:
At this point, we want a subroutine to test that a matrix is close to the identity matrix...